File: Week-Wise-Assignments/Assignment 3/Systematic Solutions/Python Programs/3_complete_linked_list.py
class Node:
    """
    Node class for single linked list
    """
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    """
    Complete Single Linked List implementation with all operations
    """
    
    def __init__(self):
        """
        Initialize empty linked list
        """
        self.head = None
    
    def createList(self):
        """
        Create an empty linked list
        """
        self.head = None
        print("Empty linked list created successfully!")
    
    # INSERT OPERATIONS
    
    def insertAtFirst(self, data):
        """
        Insert node at the beginning of the list
        """
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        print(f"Inserted {data} at the beginning")
    
    def insertAtLast(self, data):
        """
        Insert node at the end of the list
        """
        new_node = Node(data)
        
        if self.head is None:
            self.head = new_node
            print(f"Inserted {data} as the first node")
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        temp.next = new_node
        print(f"Inserted {data} at the end")
    
    def insertAtAny(self, data, position):
        """
        Insert node at any position in the list
        """
        if position < 1:
            print("Invalid position! Position should be >= 1")
            return
        
        if position == 1:
            self.insertAtFirst(data)
            return
        
        new_node = Node(data)
        temp = self.head
        
        for i in range(1, position - 1):
            if temp is None:
                break
            temp = temp.next
        
        if temp is None:
            print(f"Position {position} is beyond the list. Inserting at the end.")
            self.insertAtLast(data)
            return
        
        new_node.next = temp.next
        temp.next = new_node
        print(f"Inserted {data} at position {position}")
    
    # DELETE OPERATIONS
    
    def deleteFromFirst(self):
        """
        Delete node from the beginning of the list
        """
        if self.head is None:
            print("List is empty! Cannot delete from first.")
            return None
        
        deleted_data = self.head.data
        self.head = self.head.next
        print(f"Deleted {deleted_data} from the beginning")
        return deleted_data
    
    def deleteFromLast(self):
        """
        Delete node from the end of the list
        """
        if self.head is None:
            print("List is empty! Cannot delete from last.")
            return None
        
        if self.head.next is None:
            deleted_data = self.head.data
            self.head = None
            print(f"Deleted {deleted_data} from the end")
            return deleted_data
        
        temp = self.head
        while temp.next.next is not None:
            temp = temp.next
        
        deleted_data = temp.next.data
        temp.next = None
        print(f"Deleted {deleted_data} from the end")
        return deleted_data
    
    def deleteFromAny(self, position):
        """
        Delete node from any position in the list
        """
        if self.head is None:
            print(f"List is empty! Cannot delete from position {position}.")
            return None
        
        if position < 1:
            print("Invalid position! Position should be >= 1")
            return None
        
        if position == 1:
            return self.deleteFromFirst()
        
        temp = self.head
        for i in range(1, position - 1):
            if temp is None or temp.next is None:
                break
            temp = temp.next
        
        if temp is None or temp.next is None:
            print(f"Position {position} is beyond the list!")
            return None
        
        deleted_data = temp.next.data
        temp.next = temp.next.next
        print(f"Deleted {deleted_data} from position {position}")
        return deleted_data
    
    def deleteByValue(self, value):
        """
        Delete first occurrence of a value
        """
        if self.head is None:
            print(f"List is empty! Cannot delete value {value}.")
            return False
        
        if self.head.data == value:
            self.head = self.head.next
            print(f"Deleted value {value} from the beginning")
            return True
        
        temp = self.head
        while temp.next is not None:
            if temp.next.data == value:
                temp.next = temp.next.next
                print(f"Deleted value {value} from the list")
                return True
            temp = temp.next
        
        print(f"Value {value} not found in the list")
        return False
    
    # UTILITY OPERATIONS
    
    def displayList(self):
        """
        Display the linked list
        """
        if self.head is None:
            print("List is empty!")
            return
        
        print("Linked List: ", end="")
        temp = self.head
        while temp is not None:
            print(temp.data, end="")
            if temp.next is not None:
                print(" -> ", end="")
            temp = temp.next
        print(" -> None")
    
    def countNodes(self):
        """
        Count the number of nodes in the list
        """
        count = 0
        temp = self.head
        while temp is not None:
            count += 1
            temp = temp.next
        return count
    
    def isEmpty(self):
        """
        Check if the list is empty
        """
        return self.head is None
    
    def search(self, value):
        """
        Search for a value in the list
        """
        temp = self.head
        position = 1
        
        while temp is not None:
            if temp.data == value:
                return position
            temp = temp.next
            position += 1
        
        return -1
    
    def getNodeAtPosition(self, position):
        """
        Get data at specific position
        """
        if position < 1:
            return None
        
        temp = self.head
        for i in range(1, position):
            if temp is None:
                return None
            temp = temp.next
        
        return temp.data if temp else None
    
    def reverse(self):
        """
        Reverse the linked list
        """
        prev = None
        current = self.head
        
        while current is not None:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        self.head = prev
        print("List reversed successfully")
    
    def clear(self):
        """
        Clear the entire list
        """
        self.head = None
        print("List cleared successfully")
    
    def toList(self):
        """
        Convert linked list to Python list
        """
        result = []
        temp = self.head
        while temp is not None:
            result.append(temp.data)
            temp = temp.next
        return result
    
    def fromList(self, data_list):
        """
        Create linked list from Python list
        """
        self.clear()
        for data in data_list:
            self.insertAtLast(data)
        print(f"Created linked list from list: {data_list}")

def display_menu():
    """
    Display the menu options
    """
    print("\n=== Complete Single Linked List Operations ===")
    print("1.  Insert at First")
    print("2.  Insert at Last")
    print("3.  Insert at Any Position")
    print("4.  Delete from First")
    print("5.  Delete from Last")
    print("6.  Delete from Any Position")
    print("7.  Delete by Value")
    print("8.  Display List")
    print("9.  Count Nodes")
    print("10. Search Value")
    print("11. Get Node at Position")
    print("12. Reverse List")
    print("13. Clear List")
    print("14. Convert to Python List")
    print("15. Create from Python List")
    print("16. Exit")
    print("Enter your choice: ", end="")

def main():
    """
    Main function with interactive menu
    """
    ll = LinkedList()
    ll.createList()
    
    while True:
        display_menu()
        
        try:
            choice = int(input())
            
            if choice == 1:
                data = int(input("Enter data to insert at first: "))
                ll.insertAtFirst(data)
            
            elif choice == 2:
                data = int(input("Enter data to insert at last: "))
                ll.insertAtLast(data)
            
            elif choice == 3:
                data = int(input("Enter data to insert: "))
                position = int(input("Enter position (1-based): "))
                ll.insertAtAny(data, position)
            
            elif choice == 4:
                ll.deleteFromFirst()
            
            elif choice == 5:
                ll.deleteFromLast()
            
            elif choice == 6:
                position = int(input("Enter position to delete (1-based): "))
                ll.deleteFromAny(position)
            
            elif choice == 7:
                value = int(input("Enter value to delete: "))
                ll.deleteByValue(value)
            
            elif choice == 8:
                ll.displayList()
            
            elif choice == 9:
                print(f"Number of nodes: {ll.countNodes()}")
            
            elif choice == 10:
                value = int(input("Enter value to search: "))
                position = ll.search(value)
                if position != -1:
                    print(f"Value {value} found at position {position}")
                else:
                    print(f"Value {value} not found")
            
            elif choice == 11:
                position = int(input("Enter position to get data (1-based): "))
                data = ll.getNodeAtPosition(position)
                if data is not None:
                    print(f"Data at position {position}: {data}")
                else:
                    print(f"No data at position {position}")
            
            elif choice == 12:
                ll.reverse()
            
            elif choice == 13:
                ll.clear()
            
            elif choice == 14:
                python_list = ll.toList()
                print(f"Python list: {python_list}")
            
            elif choice == 15:
                input_str = input("Enter numbers separated by spaces: ")
                try:
                    data_list = [int(x) for x in input_str.split()]
                    ll.fromList(data_list)
                except ValueError:
                    print("Please enter valid integers separated by spaces")
            
            elif choice == 16:
                print("Exiting...")
                break
            
            else:
                print("Invalid choice! Please try again.")
        
        except ValueError:
            print("Please enter a valid number!")
        except Exception as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: Week-Wise-Assignments/Assignment 3/Systematic Solutions/C Programs/3_complete_linked_list.c
#include <stdio.h>
#include <stdlib.h>

// Node structure for single linked list
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Memory allocation failed!\n");
        return NULL;
    }
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to create an empty list
Node* createList() {
    printf("Empty linked list created successfully!\n");
    return NULL;
}

// INSERT FUNCTIONS

// Function to insert at the beginning
Node* insertAtFirst(Node* head, int data) {
    Node* newNode = createNode(data);
    if (newNode == NULL) return head;
    
    newNode->next = head;
    head = newNode;
    printf("Inserted %d at the beginning\n", data);
    return head;
}

// Function to insert at the end
Node* insertAtLast(Node* head, int data) {
    Node* newNode = createNode(data);
    if (newNode == NULL) return head;
    
    if (head == NULL) {
        printf("Inserted %d as the first node\n", data);
        return newNode;
    }
    
    Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    
    temp->next = newNode;
    printf("Inserted %d at the end\n", data);
    return head;
}

// Function to insert at any position
Node* insertAtAny(Node* head, int data, int position) {
    if (position < 1) {
        printf("Invalid position! Position should be >= 1\n");
        return head;
    }
    
    if (position == 1) {
        return insertAtFirst(head, data);
    }
    
    Node* newNode = createNode(data);
    if (newNode == NULL) return head;
    
    Node* temp = head;
    for (int i = 1; i < position - 1 && temp != NULL; i++) {
        temp = temp->next;
    }
    
    if (temp == NULL) {
        printf("Position %d is beyond the list. Inserting at the end.\n", position);
        free(newNode);
        return insertAtLast(head, data);
    }
    
    newNode->next = temp->next;
    temp->next = newNode;
    printf("Inserted %d at position %d\n", data, position);
    return head;
}

// DELETE FUNCTIONS

// Function to delete from the beginning
Node* deleteFromFirst(Node* head) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from first.\n");
        return NULL;
    }
    
    Node* temp = head;
    head = head->next;
    printf("Deleted %d from the beginning\n", temp->data);
    free(temp);
    return head;
}

// Function to delete from the end
Node* deleteFromLast(Node* head) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from last.\n");
        return NULL;
    }
    
    if (head->next == NULL) {
        printf("Deleted %d from the end\n", head->data);
        free(head);
        return NULL;
    }
    
    Node* temp = head;
    while (temp->next->next != NULL) {
        temp = temp->next;
    }
    
    Node* lastNode = temp->next;
    printf("Deleted %d from the end\n", lastNode->data);
    temp->next = NULL;
    free(lastNode);
    return head;
}

// Function to delete from any position
Node* deleteFromAny(Node* head, int position) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from position %d.\n", position);
        return NULL;
    }
    
    if (position < 1) {
        printf("Invalid position! Position should be >= 1\n");
        return head;
    }
    
    if (position == 1) {
        return deleteFromFirst(head);
    }
    
    Node* temp = head;
    for (int i = 1; i < position - 1 && temp != NULL; i++) {
        temp = temp->next;
    }
    
    if (temp == NULL || temp->next == NULL) {
        printf("Position %d is beyond the list!\n", position);
        return head;
    }
    
    Node* nodeToDelete = temp->next;
    temp->next = nodeToDelete->next;
    printf("Deleted %d from position %d\n", nodeToDelete->data, position);
    free(nodeToDelete);
    return head;
}

// UTILITY FUNCTIONS

// Function to display the list
void displayList(Node* head) {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }
    
    printf("Linked List: ");
    Node* temp = head;
    while (temp != NULL) {
        printf("%d", temp->data);
        if (temp->next != NULL) {
            printf(" -> ");
        }
        temp = temp->next;
    }
    printf(" -> NULL\n");
}

// Function to count nodes in the list
int countNodes(Node* head) {
    int count = 0;
    Node* temp = head;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// Function to search for a value
int searchValue(Node* head, int value) {
    Node* temp = head;
    int position = 1;
    
    while (temp != NULL) {
        if (temp->data == value) {
            return position;
        }
        temp = temp->next;
        position++;
    }
    return -1; // Not found
}

// Function to free the entire list
void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
    printf("Memory freed successfully!\n");
}

// Function to display menu
void displayMenu() {
    printf("\n=== Single Linked List Operations ===\n");
    printf("1.  Insert at First\n");
    printf("2.  Insert at Last\n");
    printf("3.  Insert at Any Position\n");
    printf("4.  Delete from First\n");
    printf("5.  Delete from Last\n");
    printf("6.  Delete from Any Position\n");
    printf("7.  Display List\n");
    printf("8.  Count Nodes\n");
    printf("9.  Search Value\n");
    printf("10. Clear List\n");
    printf("11. Exit\n");
    printf("Enter your choice: ");
}

int main() {
    Node* head = NULL;
    int choice, data, position, searchVal, result;
    
    printf("=== Complete Single Linked List Implementation ===\n");
    
    // Create empty list
    head = createList();
    
    do {
        displayMenu();
        scanf("%d", &choice);
        
        switch (choice) {
            case 1:
                printf("Enter data to insert at first: ");
                scanf("%d", &data);
                head = insertAtFirst(head, data);
                break;
                
            case 2:
                printf("Enter data to insert at last: ");
                scanf("%d", &data);
                head = insertAtLast(head, data);
                break;
                
            case 3:
                printf("Enter data to insert: ");
                scanf("%d", &data);
                printf("Enter position (1-based): ");
                scanf("%d", &position);
                head = insertAtAny(head, data, position);
                break;
                
            case 4:
                head = deleteFromFirst(head);
                break;
                
            case 5:
                head = deleteFromLast(head);
                break;
                
            case 6:
                printf("Enter position to delete (1-based): ");
                scanf("%d", &position);
                head = deleteFromAny(head, position);
                break;
                
            case 7:
                displayList(head);
                break;
                
            case 8:
                printf("Number of nodes: %d\n", countNodes(head));
                break;
                
            case 9:
                printf("Enter value to search: ");
                scanf("%d", &searchVal);
                result = searchValue(head, searchVal);
                if (result != -1) {
                    printf("Value %d found at position %d\n", searchVal, result);
                } else {
                    printf("Value %d not found in the list\n", searchVal);
                }
                break;
                
            case 10:
                freeList(head);
                head = NULL;
                printf("List cleared successfully!\n");
                break;
                
            case 11:
                printf("Exiting...\n");
                freeList(head);
                break;
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    } while (choice != 11);
    
    return 0;
}